### PR TITLE
[HOPSWORKS-999] Airflow recovery

### PR DESCRIPTION
**Pull request by kouzant**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [ ] Adds tests for the submitted changes (for bug fixes & features)
- [x] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**
https://logicalclocks.atlassian.net/browse/HOPSWORKS-999

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New feature

* **What is the new behavior (if this is a feature change)?**
Airflow security material will survive from Hopsworks restart

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java [hunk 1]
```diff
@@ -132,19 +148,150 @@ public void init() throws RuntimeException {
...
+        token = FileUtils.readFileToString(tokenFile.toFile(), Charset.defaultCharset());
+        DecodedJWT decoded = jwtController.verifyToken(token, settings.getJWTIssuer());
+        airflowJWT = new AirflowJWT(user.getUsername(), project.getId(), project.getName(),
+            date2LocalDateTime(decoded.getExpiresAt()), user.getUid());
+        airflowJWT.tokenFile = tokenFile;
+        airflowJWT.token = token;
+        LOG.log(Level.FINE, "Successfully read existing JWT from local filesystem for " + materialIdentifier);
+      } catch (IOException | JWTException | JWTDecodeException ex) {
+        // Could not read token or token has expired, create new
+        //Antonis: Not very good that audience is hardcoded, but it is not accessible from hopsworks-common
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java]: @kouzant replace with `TODO(Antonis):`

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java [hunk 2]
```diff
@@ -132,19 +148,150 @@ public void init() throws RuntimeException {
...
+        } catch (IOException ex1) {
+          // Managed to create token but failed to write
+          LOG.log(Level.WARNING, "Could not write to local FS new JWT for recovered material " + materialIdentifier
+            + ". We will invalidate it and won't renew it.", ex1);
+          if (token != null) {
+            try {
+              LOG.log(Level.FINE, "Failed to write JWT for " + materialIdentifier + ". Invalidating it...");
+              jwtController.invalidate(token);
+            } catch (InvalidationException ex2) {
+              // Not much we can do about it
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java]: @kouzant maybe log at `FINEST`?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java [hunk 3]
```diff
@@ -132,19 +148,150 @@ public void init() throws RuntimeException {
...
+      recover();
+    } catch (Exception ex) {
+      LOG.log(Level.WARNING, "Failed to recover material for Airflow sessions", ex);
+    }
+    
     long interval = Math.max(3000L, settings.getJWTExpLeewaySec() * 1000 / 2);
     timerService.createIntervalTimer(10L, interval, new TimerConfig("Airflow JWT renewal", false));
   }
   
+  private void recover() {
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java]: @kouzant add some comments on the workflow of this method. It's not short enough to be self-explained.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java [hunk 4]
```diff
@@ -132,19 +148,150 @@ public void init() throws RuntimeException {
...
+  @AccessTimeout(value = 5, unit = TimeUnit.SECONDS)
+  public void onProjectRemoval(Project project) throws IOException {
+    FileUtils.deleteDirectory(getProjectDagDirectory(project.getId()).toFile());
+    // Airflow material will be deleted from the database by the foreign key constraint
+    // Monitor will reap all material that do not exist in the database
+  }
+  
   @Lock(LockType.READ)
   @AccessTimeout(value = 1, unit = TimeUnit.SECONDS)
   public void prepareSecurityMaterial(Users user, Project project, String[] audience) throws AirflowException {
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java]: @kouzant add some description on what's the purpose of this timer.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/project/ProjectController.java [hunk 5]
```diff
@@ -1275,6 +1296,36 @@ public void removeProject(String userMail, int projectId, String sessionId) thro
...
+        List<ProjectTeam> reconstructedProjectTeam = new ArrayList<>();
+        try {
+          for (HdfsUsers hdfsUser : hdfsUsersController.getAllProjectHdfsUsers(projectName)) {
+            Users foundUser = userFacade.findByUsername(hdfsUser.getUsername());
+            if (foundUser != null) {
+              reconstructedProjectTeam.add(new ProjectTeam(toDeleteProject, foundUser));
+            }
+          }
+        } catch (Exception ex) {
+          // NOOP
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/project/ProjectController.java]: @kouzant when is this exception thrown?

Comment (kouzant) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/project/ProjectController.java]: This method is a *forced* project deletion. It should delete as much as possible without failing. In that line, an exception might be thrown from the facades.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java (hunk 2 above)
Comment (kouzant) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/airflow/AirflowManager.java]: I put some to FINEST

**Pull request merged**